apis/oam/v1alpha2/core_types.go

```diff
@@ -0,0 +1,375 @@
+type WorkloadDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []WorkloadDefinition `json:"items"`
+}
+
+// A TraitDefinitionSpec defines the desired state of a TraitDefinition.
+type TraitDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this trait kind.
+	Reference DefinitionReference `json:"definitionRef"`
+
+	// AppliesToWorkloads specifies the list of workload kinds this trait
+	// applies to. Workload kinds are specified in kind.group/version format,
+	// e.g. server.core.oam.dev/v1alpha2. Traits that omit this field apply to
+	// all workload kinds.
+	// +optional
+	AppliesToWorkloads []string `json:"appliesToWorkloads,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// A TraitDefinition registers a kind of Kubernetes custom resource as a valid
+// OAM trait kind by referencing its CustomResourceDefinition. The CRD is used
+// to validate the schema of the trait when it is embedded in an OAM
+// ApplicationConfiguration.
+// +kubebuilder:printcolumn:JSONPath=".spec.definitionRef.name",name=DEFINITION-NAME,type=string
+type TraitDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec TraitDefinitionSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// TraitDefinitionList contains a list of TraitDefinition.
+type TraitDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []TraitDefinition `json:"items"`
+}
+
+// A ScopeDefinitionSpec defines the desired state of a ScopeDefinition.
+type ScopeDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this scope kind.
+	Reference DefinitionReference `json:"definitionRef"`
+
+	// AllowComponentOverlap specifies whether an OAM component may exist in
+	// multiple instances of this kind of scope.
+	AllowComponentOverlap bool `json:"allowComponentOverlap"`
+}
+
+// +kubebuilder:object:root=true
+
+// A ScopeDefinition registers a kind of Kubernetes custom resource as a valid
+// OAM scope kind by referencing its CustomResourceDefinition. The CRD is used
+// to validate the schema of the scope when it is embedded in an OAM
+// ApplicationConfiguration.
+// +kubebuilder:printcolumn:JSONPath=".spec.definitionRef.name",name=DEFINITION-NAME,type=string
+type ScopeDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ScopeDefinitionSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ScopeDefinitionList contains a list of ScopeDefinition.
+type ScopeDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ScopeDefinition `json:"items"`
+}
+
+// A ComponentParameter defines a configurable parameter of a component.
+type ComponentParameter struct {
```

Comment:
I think we should remove `default` because you can set a default by simply including the value(s) the parameter would set in `spec.workload`. The `description` is still valuable though - will fix that.

Reply:
It appears both `default` and `description` are removed in https://github.com/oam-dev/spec/pull/316 so will need to update there.

Reply:
Updated, I added back `description` only.

Reply:
make sense